flag.go

```diff
@@ -0,0 +1,192 @@
+package ldclient
+
+import (
+	"crypto/sha1"
+	"encoding/hex"
+	"io"
+	"reflect"
+	"strconv"
+)
+
+type FeatureFlag struct {
+	Name         string      `json:"name" bson:"name"`
+	Key          string      `json:"key" bson:"key"`
+	Version      int         `json:"version" bson:"version"`
+	On           bool        `json:"on" bson:"on"`
+	Salt         string      `json:"salt" bson:"salt"`
+	Sel          string      `json:"sel" bson:"sel"`
+	Targets      []Target    `json:"targets" bson:"targets"`
+	Rules        []Rule      `json:"rules" bson:"rules"`
+	Fallthrough  Rule        `json:"fallthrough" bson:"fallthrough"`
+	OffVariation interface{} `json:"offVariation" bson:"offVariation"`
+	Archived     bool        `json:"archived" bson:"archived"`
+}
+
+// Expresses a set of AND-ed matching conditions for a user, along with
+// either the fixed variation or percent rollout to serve if the conditions
+// match.
+// Invariant: one of the variation or rollout must be non-nil.
+type Rule struct {
+	Clauses   []Clause    `json:"clauses,omitempty" bson:"clauses,omitempty"`
+	Variation interface{} `json:"variation,omitempty" bson:"variation,omitempty"`
+	Rollout   *Rollout    `json:"rollout,omitempty" bson:"rollout,omitempty"`
+}
+
+type Rollout struct {
+	Variations []WeightedVariation `json:"variations,omitempty" bson:"variations"`
+	BucketBy   *string             `json:"bucketBy,omitempty" bson:"bucketBy,omitempty"`
+}
+
+type Clause struct {
+	Attribute string        `json:"attribute" bson:"attribute"`
+	Op        Operator      `json:"op" bson:"op"`
+	Values    []interface{} `json:"values" bson:"values"` // An array, interpreted as an OR of values
+	Negate    bool          `json:"negate" bson:"negate"`
+}
+
+type WeightedVariation struct {
+	Variation interface{} `json:"variation" bson:"variation"`
+	Weight    int         `json:"weight" bson:"weight"` // Ranges from 0 to 100000
+}
+
+type Target struct {
+	Values    []string    `json:"values" bson:"values"`
+	Variation interface{} `json:"variation" bson:"variation"`
+}
+
+// An explanation is either a target or a rule
+type Explanation struct {
+	Kind string `json:"kind" bson:"kind"`
+	*Target
+	*Rule
+}
+
+func bucketUser(user User, key, attr, salt string) float32 {
+
+	uValue, pass := user.valueOf(attr)
+
+	if idHash, ok := uValue.(string); pass || !ok {
+		return 0
+	} else {
+		if user.Secondary != nil {
+			idHash = idHash + "." + *user.Secondary
+		}
+
+		h := sha1.New()
+		io.WriteString(h, key+"."+salt+"."+idHash)
+		hash := hex.EncodeToString(h.Sum(nil))[:15]
+
+		intVal, _ := strconv.ParseInt(hash, 16, 64)
+
+		bucket := float32(intVal) / long_scale
+
+		return bucket
+	}
+}
+
+func (f FeatureFlag) EvaluateExplain(user User) (interface{}, *Explanation) {
```

Comment:
exported method FeatureFlag.EvaluateExplain should have comment or be unexported